File: Bronze/2019/December/02_where_am_i/WhereAmI.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.HashSet;

/**
 * @author Red
 * @version 1.0
 */
public class WhereAmI {
    static HashSet<String> mailBox = new HashSet<String>();
    static int n;

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("whereami.in"));
        PrintWriter pw = new PrintWriter(new FileWriter("whereami.out"));
        n = Integer.parseInt(r.readLine());
        String boxes = r.readLine();
        int ans = n;
        boolean ok;
        // 依次截取不同长度的子串，判断是否有重复
        for (int scope = 1; scope <= n; scope++) {
            ok = true;
            for (int i = 0; i <= n - scope; i++) {
                String sub = boxes.substring(i, i + scope);
                if (!mailBox.contains(sub)) {
                    mailBox.add(sub);
                } else {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                ans = scope;
                break;
            }
        }
        pw.println(ans);
        pw.close();
    }
}

File: Bronze/2019/December/02_where_am_i/where_am_i.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>

using namespace std;

int main() {
    ifstream fin("whereami.in");
    ofstream fout("whereami.out");

    int n;
    fin >> n;

    string boxes;
    fin >> boxes;

    unordered_set<string> mailBox;
    int ans = n;
    bool ok;
    // 依次截取不同长度的子串，判断是否有重复
    for (int scope = 1; scope <= n; scope++) {
        ok = true;
        for (int i = 0; i <= n - scope; i++) {
            string sub = boxes.substr(i, scope);
            if (mailBox.find(sub) == mailBox.end()) {
                mailBox.insert(sub);
            } else {
                ok = false;
                break;
            }
        }
        if (ok) {
            ans = scope;
            break;
        }
    }

    fout << ans << endl;

    fin.close();
    fout.close();

    return 0;
}
